File: elixir/lib/2023/09_mirage_maintenance/aoc202309.ex
defmodule AOC2023.Day09 do
  @moduledoc """
  Advent of Code 2023, day 9: Mirage Maintenance.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input) do
    puzzle_input
    |> String.split("\n", trim: true)
    |> Enum.map(fn line -> String.split(line) |> Enum.map(&String.to_integer/1) end)
  end

  @doc """
  Solve part 1.
  """
  def part1(data) do
    data |> Enum.map(fn numbers -> numbers |> Enum.reverse() |> extrapolate() end) |> Enum.sum()
  end

  @doc """
  Solve part 2.
  """
  def part2(data) do
    data |> Enum.map(&extrapolate/1) |> Enum.sum()
  end

  @doc """
  Extrapolate sequence to find the previous element.

  Extrapolate to the left, since it's easier to work with the head of a list in
  Elixir.

  ## Example:

      iex> extrapolate([0, 2, 5, 9, 14])
      -1
  """
  def extrapolate(numbers), do: extrapolate(numbers, numbers |> hd())

  def extrapolate(numbers, prev_value) do
    if Enum.all?(numbers, &(&1 == 0)) do
      prev_value
    else
      diff =
        numbers
        |> Enum.chunk_every(2, 1, :discard)
        |> Enum.map(fn [second, first] -> second - first end)

      extrapolate(diff, prev_value + (diff |> hd()))
    end
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/2023/09_mirage_maintenance/aoc202309.py
"""AoC 9, 2023: Mirage Maintenance."""

# Standard library imports
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input."""
    return [[int(num) for num in line.split()] for line in puzzle_input.split("\n")]


def part1(data):
    """Solve part 1."""
    return sum(extrapolate(numbers) for numbers in data)


def part2(data):
    """Solve part 2."""
    return sum(extrapolate(numbers[::-1]) for numbers in data)


def extrapolate(numbers):
    """Extrapolate a number series to find the next number in the series.

    ## Example:

    >>> extrapolate([1, 3, 6, 10, 15, 21])
    28
    """
    last_diff = [numbers[-1]]
    while any(diff != 0 for diff in numbers):
        numbers = [second - first for first, second in zip(numbers, numbers[1:])]
        last_diff.append(numbers[-1])

    return sum(last_diff)


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))
